### DISCUSSION TITLE
How to dynamically issue triggers with DI?

### DISCUSSION THREAD

**Discussion by nikriaz** [quartznet/quartznet#1907] [Q&A]
It maybe simple or silly question but how to issue triggers dynamically if the job scheduler is registered as DI hosted service? According to the [official tutorial](https://www.quartz-scheduler.net/documentation/quartz-3.x/packages/microsoft-di-integration.html#di-aware-job-factories),  triggers are added in Startup class. What if I need them later and what if need a lot of them?

I need to ask public API, and if a data is not there, re-request with gradually increasing time within 24 hours. So, my plan to issue one-time trigger for every failed request to API with Quartz with calculated next-time fire. There will be hundreds of such triggers a day. Is it okay plan? What would be the best practices for such typical task?

**Comment by lahma** [comment]
If you want to do dynamic maintenance, inject a `ISchedulerFactory` to your logic layer and call `IScheduler scheduler = await schedulerFactory.GetScheduler();`. Now you have the scheduler API to do CRUD operation to triggers and jobs. You can also do further scheduling from job side if needed (context has reference to `IScheduler`).

**Reply by nikriaz** [reply]
Thank you! 

I should admit I don't fully understand the nature of triggers. Are they like database entry or instantiated class with task, kind of fire-and-forget? I even used to think they are more like hardcoded static function (like in the sample for Startup class), but seem they are not.

**Reply by lahma** [reply]
Maybe think trigger as a stateful rule. 

1.  Quartz asks the trigger for the next fire time after this moment (`DateTime.UtcNow`), it will reply "at midnight" 
2. Quartz  will then poll and check for all triggers and sees that it's now midnight and this trigger (and maybe some others) should run
3. Quartz check what job is targeted by the trigger and schedules the job to run - job is executed
4. Trigger can now persist information like (I've run 10 times already), in-memory or database
5. Next time trigger is queries for info (see step 1) it can reply "again at this new day's midnight" or decide to say "never again" as I'm only allowed to run at most 10 times, or I'm allowed to run until year 2024 and then I should stop

So job is separated entity from trigger. It can live without a trigger if configured so (durable). There can be 0-N triggers targeting same job. Triggers advice when to execute some piece of code (job).

**Reply by nikriaz** [reply]
Awesome, this is very useful, thank you! Will play a bit with concrete implementations to get a feeling how it works.
